### DISCUSSION TITLE
Can recipe parameters be made mandatory?

### DISCUSSION THREAD

**Discussion by gabro** [vanilla-extract-css/vanilla-extract#468] [Q&A]
In my (admittedly limited) experience with the recipe API, I find myself wanting the parameters of a recipe to be mandatory.

E.g.

```ts
// Button.css.ts

export const buttonRecipe = recipe({
  variants: {
    size: {
      large: { height: "48px" },
      small: { height: "32px" }
    }
  }
})

// Button.tsx
import { buttonRecipe } from "./Button.css";

export function Button(props: { size: "large" | "small" }) {
  return <button className={buttonReceipe()} /> // this is a bug, I forgot to set 'size' and TypeScript is not complaining
}
```

I would expect the type of a parameter to be optional only when a `defaultVariant` is provided.

So two questions here:

- what's the rationale behind making `recipe` parameters optional?
- is there a way to make this stricter?

Thanks!

**Comment by markdalgleish** [comment]
Just had a go at implementing this: https://github.com/seek-oss/vanilla-extract/pull/469

**Reply by gabro** [reply]
That's great! Just wondering, why is a variant optional by default? Wouldn't it be safer to have it mandatory by default and reverse the API? Is it only to avoid breaking changes or is it something more deliberate in the design?

**Reply by markdalgleish** [reply]
We were following Stitches' lead here, and I'm trying to avoid a breaking change. Worth considering though.

**Reply by gabro** [reply]
I totally understand the desire not to introduce breaking changes, so no arguments there.

My 2 cents about the default (in case it's something you want to consider changing anyway): when the default is optional, forgetting to pass a value or to use the propose `requiredVariants` option can result in a silent bug.
Conversely, if parameters are mandatory by default, forgetting to pass a value will result in a type error, at which point one may decide to use an api like `optionalVariants` (the inverse of `requiredVariants`) to relax the constraint.
This way it would be much harder to accidentally introduce a bug, and it feels more on brand with the general philosophy of this library (which favors strictness)

**Reply by markdalgleish** [reply]
Yep, I agree that it feels more on brand for us to be stricter by default. Might chat with the rest of the team about alternative APIs.

**Reply by gabro** [reply]
Thanks for considering it.
Incidentally, I found that a few days ago the same question was raised for stitches https://github.com/modulz/stitches/discussions/861

**Reply by kylemh** [reply]
Hilariously fun to see identical conclusions reached in similar libraries with none of the same people in the convo ❤️
